src/cdk-experimental/table-scroll-container/table-scroll-container.ts

```diff
@@ -0,0 +1,136 @@
+import {Directive, ElementRef, Inject, OnDestroy, Optional} from '@angular/core';
+import {DOCUMENT} from '@angular/common';
+import {Directionality} from '@angular/cdk/bidi';
+import {StickyPositioningListener} from '@angular/cdk/table';
+
+let nextId = 0;
+
+type Orientation = 'horizontal' | 'vertical';
+
+@Directive({
+  selector: '[cdkTableScrollContainer]',
+  providers: [
+    {provide: StickyPositioningListener, useExisting: CdkTableScrollContainer},
+  ],
+})
+export class CdkTableScrollContainer implements StickyPositioningListener,
+    OnDestroy {
+  protected readonly CLASS_NAME = 'cdk-table-scroll-container';
+
+  private readonly _document: Document;
+  private readonly _element: HTMLElement;
+  private readonly _uniqueClassName: string;
+  private readonly _ruleIndexes = new Map<Orientation, number>();
+
+  private _styleElement?: HTMLStyleElement;
+  private _indexSequence = 0;
+  private _startSizes: (number|null|undefined)[] = [];
```

Comment:
What's the context having both `null` and `undefined` for these signatures? Also you're repeating it in a bunch of places, maybe it can be moved out into a `type`?

Reply:
It makes the implementation in StickyStyler a bit simpler, and this has no reason to care which one it is. I can pull it out to a type.